File: Sources/WattsonCore/Quotas/Poller.swift
// Quotas/Poller.swift — 多 provider 订阅额度轮询：单飞 + TTL（5 分钟）；
// 各 provider 相互隔离，任一失败不影响其它；全部失败/无凭据只体现在各自
// account 的 unavailableReason 里。
import Foundation

let PROVIDER_ORDER: [(QuotaKind, String)] = [
    (.glm, "GLM Coding Plan"),
    (.codex, "Codex (ChatGPT)"),
    (.claude, "Claude Code"),
    (.cursor, "Cursor"),
    (.workbuddy, "WorkBuddy"),
    (.trae, "Trae"),
    (.kimi, "Kimi"),
    (.gemini, "Gemini"),
    (.grok, "Grok"),
    (.zed, "Zed"),
    (.kiro, "Kiro"),
    (.codebuff, "Codebuff"),
    (.factory, "Factory"),
    (.copilot, "Copilot"),
    (.openrouter, "OpenRouter"),
    (.minimax, "MiniMax"),
]

public final class QuotaPoller: @unchecked Sendable {
    private let lock = NSLock()
    private var state: QuotaSnapshot
    private var inflight: Task<QuotaSnapshot, Never>?
    private var lastFetchAt = -Double.infinity

    private let home: String
    private let env: [String: String]
    private let doFetch: FetchLike
    private let now: @Sendable () -> Double

    public init(home: String? = nil, env: [String: String]? = nil,
                fetchImpl: FetchLike? = nil, now: (@Sendable () -> Double)? = nil) {
        self.home = home ?? homePath()
        self.env = env ?? ProcessInfo.processInfo.environment
        self.doFetch = fetchImpl ?? makeHTTPFetch()
        self.now = now ?? { Date.nowMs() }
        // loading 占位：首拉完成前 UI 有确定形态
        self.state = QuotaSnapshot(accounts: PROVIDER_ORDER.map {
            QuotaAccount(kind: $0.0, label: $0.1, available: false, unavailableReason: "loading",
                         error: nil, planName: nil, windows: [], fetchedAt: 0, lastSuccessAt: nil)
        })
    }

    /// 最近一轮快照（同步读取：HTTP 服务器与 UI 都用它）
    public var snapshotValue: QuotaSnapshot {
        lock.lock(); defer { lock.unlock() }
        return state
    }

    /// 拉一轮全部 provider（带 TTL）；绝不抛错
    public func current() async -> QuotaSnapshot {
        lock.lock()
        if let t = inflight {
            lock.unlock()
            return await t.value
        }
        if now() - lastFetchAt < PLAN_TTL_MS {
            lock.unlock()
            return snapshotValue
        }
        let task = Task<QuotaSnapshot, Never> { await self.runRound() }
        inflight = task
        lock.unlock()
        return await task.value
    }

    private func runRound() async -> QuotaSnapshot {
        let n = now()
        // 各 provider 并发拉取，按 PROVIDER_ORDER 顺序回填
        var results = [QuotaAccount?](repeating: nil, count: PROVIDER_ORDER.count)
        await withTaskGroup(of: (Int, QuotaAccount).self) { group in
            for (idx, spec) in PROVIDER_ORDER.enumerated() {
                group.addTask {
                    let account = await self.fetchOne(kind: spec.0, label: spec.1, now: n)
                    return (idx, account)
                }
            }
            for await (idx, account) in group {
                results[idx] = account
            }
        }
        let snapshot = QuotaSnapshot(accounts: results.enumerated().map { idx, account in
            account ?? loadingAccount(PROVIDER_ORDER[idx].0, PROVIDER_ORDER[idx].1, n)
        })
        lock.lock()
        state = snapshot
        lastFetchAt = now()
        inflight = nil
        lock.unlock()
        return snapshot
    }

    private func fetchOne(kind: QuotaKind, label: String, now: Double) async -> QuotaAccount {
        do {
            return try await fetchProvider(kind, home, env, doFetch, now)
        } catch {
            // provider 隔离：任何未捕获异常都降级成该账号的 error 快照
            return QuotaAccount(kind: kind, label: label, available: false,
                                unavailableReason: "error", error: errText(error),
                                planName: nil, windows: [], fetchedAt: now, lastSuccessAt: nil)
        }
    }
}

func loadingAccount(_ kind: QuotaKind, _ label: String, _ now: Double) -> QuotaAccount {
    QuotaAccount(kind: kind, label: label, available: false, unavailableReason: "loading",
                 error: nil, planName: nil, windows: [], fetchedAt: now, lastSuccessAt: nil)
}

/// 单 provider 分发：glm 走 PlanSnapshot（绝对值口径），其余各自模块
func fetchProvider(_ kind: QuotaKind, _ home: String, _ env: [String: String],
                   _ doFetch: FetchLike, _ now: Double) async throws -> QuotaAccount {
    switch kind {
    case .glm:
        guard let auth = readCodingPlanAuth(home, env) else {
            return missingAccount(.glm, "GLM Coding Plan", now)
        }
        let snap = await fetchPlanSnapshot(auth: auth, doFetch: doFetch, now: now)
        return glmToAccount(snap)
    case .codex: return await codexAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .claude: return await claudeAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .cursor: return await cursorAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .workbuddy: return await workbuddyAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .trae: return await traeAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .kimi: return await kimiAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .gemini: return await geminiAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .grok: return await grokAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .zed: return await zedAccount(home, env: env, doFetch: doFetch, now: now)
    case .kiro: return await kiroAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .codebuff: return await codebuffAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .factory: return await factoryAccount(home: home, env: env, doFetch: doFetch, now: now)
    case .copilot: return await copilotAccount(home, env: env, doFetch: doFetch, now: now)
    case .openrouter: return await openrouterAccount(home, env: env, doFetch: doFetch, now: now)
    case .minimax: return await minimaxAccount(home, env: env, doFetch: doFetch, now: now)
    }
}

File: server/src/quotas/poller.ts
// server/src/quotas/poller.ts — 多 provider 订阅额度轮询：
// 单飞 + TTL（5 分钟）；各 provider 相互隔离，任一失败不影响其它；
// 全部失败/无凭据只体现在各自 account 的 unavailableReason 里。
import { homedir } from 'node:os'
import {
  fetchPlanSnapshot, readCodingPlanAuth, PLAN_TTL_MS, type PlanSnapshot,
} from '../plan.js'
import { codexAccount } from './codex.js'
import { claudeAccount } from './claude.js'
import { cursorAccount } from './cursor.js'
import { workbuddyAccount } from './workbuddy.js'
import { traeAccount } from './trae.js'
import { kimiAccount } from './kimi.js'
import { geminiAccount } from './gemini.js'
import { grokAccount } from './grok.js'
import { zedAccount } from './zed.js'
import { kiroAccount } from './kiro.js'
import { codebuffAccount } from './codebuff.js'
import { factoryAccount } from './factory.js'
import { copilotAccount } from './copilot.js'
import { openrouterAccount } from './openrouter.js'
import { minimaxAccount } from './minimax.js'
import type { FetchLike, QuotaAccount, QuotaSnapshot } from './types.js'

/** GLM 的 PlanSnapshot（绝对值口径）→ 统一 QuotaAccount */
export function glmToAccount(s: PlanSnapshot): QuotaAccount {
  return {
    kind: 'glm',
    label: 'GLM Coding Plan',
    available: s.available,
    unavailableReason: s.unavailableReason,
    error: s.error,
    planName: s.planName,
    windows: s.windows.map(w => ({
      key: w.key,
      label: w.label,
      total: w.total,
      used: w.used,
      remaining: w.remaining,
      percentage: w.percentage,
      nextResetAt: w.nextResetAt,
    })),
    fetchedAt: s.fetchedAt,
    lastSuccessAt: s.lastSuccessAt,
  }
}

const LOADING_GLM: QuotaAccount = {
  kind: 'glm', label: 'GLM Coding Plan', available: false,
  unavailableReason: 'loading', error: null, planName: null,
  windows: [], fetchedAt: 0, lastSuccessAt: null,
}

const LOADING_ACCOUNTS: QuotaAccount[] = [
  LOADING_GLM,
  { kind: 'codex', label: 'Codex (ChatGPT)', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'claude', label: 'Claude Code', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'cursor', label: 'Cursor', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'workbuddy', label: 'WorkBuddy', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'trae', label: 'Trae', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'kimi', label: 'Kimi', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'gemini', label: 'Gemini', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'grok', label: 'Grok', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'zed', label: 'Zed', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'kiro', label: 'Kiro', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'codebuff', label: 'Codebuff', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'factory', label: 'Factory', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'copilot', label: 'Copilot', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'openrouter', label: 'OpenRouter', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
  { kind: 'minimax', label: 'MiniMax', available: false, unavailableReason: 'loading', error: null, planName: null, windows: [], fetchedAt: 0, lastSuccessAt: null },
]

export class QuotaPoller {
  snapshot: QuotaSnapshot = { accounts: LOADING_ACCOUNTS }
  private inflight: Promise<void> | null = null
  private lastFetchAt = Number.NEGATIVE_INFINITY
  private readonly home: string
  private readonly env: NodeJS.ProcessEnv
  private readonly doFetch: FetchLike
  private readonly now: () => number

  constructor(opts: { home?: string; env?: NodeJS.ProcessEnv; fetchImpl?: FetchLike; now?: () => number } = {}) {
    this.home = opts.home ?? homedir()
    this.env = opts.env ?? process.env
    this.doFetch = opts.fetchImpl ?? (defaultFetch as unknown as FetchLike)
    this.now = opts.now ?? (() => Date.now())
  }

  /** 拉一轮全部 provider（带 TTL）；绝不 reject */
  async current(): Promise<QuotaSnapshot> {
    if (this.inflight) return this.inflight.then(() => this.snapshot)
    if (this.now() - this.lastFetchAt < PLAN_TTL_MS) return this.snapshot
    this.inflight = (async () => {
      const now = this.now()
      const results = await Promise.all([
        this.fetchGlm(now),
        Promise.resolve(codexAccount(this.home, this.env, this.doFetch, now)).catch(isolate('codex', now)),
        Promise.resolve(claudeAccount(this.home, this.env, this.doFetch, now)).catch(isolate('claude', now)),
        Promise.resolve(cursorAccount(this.home, this.env, this.doFetch, now)).catch(isolate('cursor', now)),
        Promise.resolve(workbuddyAccount(this.home, this.env, this.doFetch, now)).catch(isolate('workbuddy', now)),
        Promise.resolve(traeAccount(this.home, this.env, this.doFetch, now)).catch(isolate('trae', now)),
        Promise.resolve(kimiAccount(this.home, this.env, this.doFetch, now)).catch(isolate('kimi', now)),
        Promise.resolve(geminiAccount(this.home, this.env, this.doFetch, now)).catch(isolate('gemini', now)),
        Promise.resolve(grokAccount(this.home, this.env, this.doFetch, now)).catch(isolate('grok', now)),
        Promise.resolve(zedAccount(this.home, this.env, this.doFetch, now)).catch(isolate('zed', now)),
        Promise.resolve(kiroAccount(this.home, this.env, this.doFetch, now)).catch(isolate('kiro', now)),
        Promise.resolve(codebuffAccount(this.home, this.env, this.doFetch, now)).catch(isolate('codebuff', now)),
        Promise.resolve(factoryAccount(this.home, this.env, this.doFetch, now)).catch(isolate('factory', now)),
        Promise.resolve(copilotAccount(this.home, this.env, this.doFetch, now)).catch(isolate('copilot', now)),
        Promise.resolve(openrouterAccount(this.home, this.env, this.doFetch, now)).catch(isolate('openrouter', now)),
        Promise.resolve(minimaxAccount(this.home, this.env, this.doFetch, now)).catch(isolate('minimax', now)),
      ])
      this.snapshot = { accounts: results }
    })().finally(() => {
      this.lastFetchAt = this.now()
      this.inflight = null
    })
    return this.inflight.then(() => this.snapshot)
  }

  private async fetchGlm(now: number): Promise<QuotaAccount> {
    try {
      const auth = readCodingPlanAuth(this.home, this.env)
      if (!auth) return missing('glm', 'GLM Coding Plan', now)
      const snap = await fetchPlanSnapshot(auth, this.doFetch, now)
      return glmToAccount(snap)
    } catch (err) {
      return {
        kind: 'glm', label: 'GLM Coding Plan', available: false,
        unavailableReason: 'error', error: err instanceof Error ? err.message : String(err),
        planName: null, windows: [], fetchedAt: now, lastSuccessAt: null,
      }
    }
  }
}

const missing = (kind: QuotaAccount['kind'], label: string, now: number): QuotaAccount => ({
  kind, label, available: false, unavailableReason: 'no_credentials', error: null,
  planName: null, windows: [], fetchedAt: now, lastSuccessAt: null,
})

/** provider 隔离：任何未捕获异常都降级成该账号的 error 快照 */
function isolate(kind: QuotaAccount['kind'], now: number): (err: unknown) => QuotaAccount {
  const label = LOADING_ACCOUNTS.find(a => a.kind === kind)?.label ?? kind
  return (err: unknown) => ({
    kind, label, available: false, unavailableReason: 'error',
    error: err instanceof Error ? err.message : String(err),
    planName: null, windows: [], fetchedAt: now, lastSuccessAt: null,
  })
}

const defaultFetch: FetchLike = (url, init) => fetch(url, init)
